packages/kbn-i18n/src/core.js

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import IntlMessageFormat from 'intl-messageformat';
+import memoizeIntlConstructor from 'intl-format-cache';
+
+const EN_LOCALE = 'en';
+
+const getMessageById = (messages, id) =>
+  id.split('.').reduce((val, key) => (val ? val[key] : null), messages);
+
+const hasValues = values => Object.keys(values).length > 0;
+
+const addLocaleData = localeData => {
+  if (localeData && localeData.locale) {
+    IntlMessageFormat.__addLocaleData(localeData);
+  }
+};
+
+const normalizeLocale = (locale = '') => locale.toLowerCase().replace('_', '-');
+
+const showError = error => {
```

Comment:
__nit:__ no need in `showError`, just throw errors directly.

Reply:
**question**: This is not a criticism. I'm just wondering why we have a few checks for `process.env.NODE_ENV !== 'production'`. Why do we have to make sure it's not production?

Reply:
The original idea was to show errors only in dev environment and to use fallbacks if something went wrong. If we will throw errors directly, user will get empty labels for unsuccessful translations instead of `defaultMessage`s or translations ids (as a final fallback).

Reply:
Let's see, we have 9 cases when we call `showError` right now:
1. `addMessages`: `[I18n] A "locale" must be provided to add messages.` - looks like unrecoverable error for both dev and prod, developer or lib bug - I believe we should just throw here
2. `setLocale`: `[I18n] A "locale" must be not non-empty string.` - same here
3. `setDefaultLocale`: `[I18n] A "locale" must be not non-empty string.` - same here
4. `setFormats`: `[I18n] A "formats" must be not non-empty object.` - same here
5. `translate`: `[I18n] An `id` must be provided to translate a message.` - same here
6. `translate`: `Error formatting message...` - it's the error that __is thrown__ by either `i18n.getMessageFormat(...)` or `msg.format(...)` when we format message - in what cases these methods can throw? Is it expected/normal program flow or not? If not, then we also should just re-throw the error in both prod and dev
7. `translate`: `Error formatting default message...` - same as 6., but for default message
8. `translate`: `[I18n] Missing message:` - in this case we have neither default message nor message for the specified locale - sounds like a bug to me as well, since we should have caught the absence of default message with our tools, so I'd throw here as well
9. `translate`: `[I18n] Cannot format message` - seems to be very similar to 7. and 8.

So I see only one case where I'd probably log something - it's when locale is not default and we don't have message for this locale, only default message (that we should always have if there are no bugs in our tools). And it would be `warn`, `info` or even `debug`, not `error`. The problem is that we should log this with `kbnServer.log` on server and with `console.*` on client and I don't have good ideas on how to do that properly right now, so I'd propose just ignore it on the first iteration.

>  translations ids (as a final fallback).

I don't think we should have this final fallback since `defaultMessage` is mandatory. 

Thoughts? Am I missing something?

Reply:
Agree, I replaced `showError` method with `throw new Error`